### DISCUSSION TITLE
Document.extract_image seems to extract image png mirrored in some cases

### DISCUSSION THREAD

**Discussion by p1-dta** [pymupdf/PyMuPDF#1217] [Looking for help] [answered]
Hi,

I have a pdf with only a png in it, but when I get try to extract this png, it's mirrored.

Document.extract_image return a dictionary with, for the value "image", the image but mirrored.
This happen only for one file, works fine for many more.

I have no clues, neither do my colleagues.
I don't know how to prevent it or how to detect a mirror to mirror it again.

**Comment by JorjMcKie** [comment]
Guilty is the PDF creator, who decided to do this!
You have the option to extract the "transformation matrix" that was used to make an image fit into its bbox on the page (scaling, mirroring, rotating).
Analyzing this matrix will give you the desired information.
Read the documentation [here](https://pymupdf.readthedocs.io/en/latest/app4.html) for details.
This code snippet may help with such a matrix analysis:
```python
import fitz

msg = ""
m = fitz.Matrix()
if m.b * m.c == 0:  # means rotation by 0, 180, or flippings
    if m.a * m.d > 0:  # same sign -> no flippings
        if m.a < 0:  # so both, a and d are negative!
            msg = "rot 180"
        else:
            msg = "nothing"
    else:  # we have a flip
        if m.a < 0:  # horizontal flip
            msg = "left-right"
        else:
            msg = "up-down"
elif abs(m.b) == abs(m.c):
    if m.b * m.c < 0:
        if m.b > 0:
            msg = "rot 90"
        else:
            msg = "rot 270"
    else:
        if m.b > 0:
            msg = "up-down, rot 90"
        else:
            msg = "rot 90, updown"
else:
    error = True

if error:
    print("unsupported matrix")
else:
    print(msg)
```

**Reply by p1-dta** [reply]
Thanks, I mark the question as solved if this solve my issue.

**Reply by p1-dta** [reply]
I get the matrix with `_, mat = page.get_image_bbox(page.get_images(full=True)[0], transform=True)` but `mat ==  Matrix(850.0, -0.0, -0.0, 1169.0, -0.0, 1169.0)` and the matrix analyzer return "nothing".

I compared 2 images, one flipped up-down (should be d = -1), one normal, and here is the respective matrix :
```
flipped up-down: Matrix(850.0, -0.0, -0.0, 1169.0, -0.0, 1169.0)
normal: Matrix(402.5350036621094, -0.0, -0.0, 841.9999389648438, 96.23249816894531, -0.0)
```

I can't tell the difference without looking the result with my eyes.

I checked:

- https://pymupdf.readthedocs.io/en/latest/page.html#Page.get_image_bbox
- https://pymupdf.readthedocs.io/en/latest/document.html#Document.get_page_images
- https://pymupdf.readthedocs.io/en/latest/matrix.html#examples
and [a, b, c, d, e, f] attributes.

**Comment by JorjMcKie** [comment]
What do the EXIF data of the images say?

**Reply by p1-dta** [reply]
doc.extract_image(xref) doesn't contain the EXIF value.

**Reply by p1-dta** [reply]
I'm looking at https://github.com/pymupdf/PyMuPDF/issues/385 but no solution found yet

**Reply by JorjMcKie** [reply]
Then I need an example for unexpected behaviour

**Reply by p1-dta** [reply]
The unexpected behavior is that the image is flipped up-down but `page.get_image_bbox(image, transform=True)` return a mat with a d > 0.

**Reply by p1-dta** [reply]
@JorjMcKie What information can I give you to help understanding what happens ?

**Reply by JorjMcKie** [reply]
I need a file example.

**Reply by p1-dta** [reply]
Unfortunately, I can't share this document. Is there anything else I can do ?

**Reply by JorjMcKie** [reply]
No, I did understand your issue, but the only way for further investigation is looking at the file itself.

**Reply by p1-dta** [reply]
I'll ask if it's possible, maybe in a few days. Thanks for your help.

**Reply by p1-dta** [reply]
@JorjMcKie I have an update. I can share the file with you. Do you want me to send it to you on your mailbox ?

**Comment by JorjMcKie** [comment, marked as answer]
Yes please do so.
Yesterday I found an error in computing the transformation matrix via `get_image_bbox()`. This may affect you. You can try `get_image_rects()` instead. This is always correct. It returns a list of tuples `(rect, matrix)`. One tuple for each occurrence of the same image on the page. You probably will find the correct answer with this alternative.
Von meinem iPhone gesendet

Am 6/9/21 um 05:08 schrieb Dorian Turba ***@***.***>:

﻿

@JorjMcKie<https://github.com/JorjMcKie> I have an update. I can share the file with you. Do you want me to send it to you on your mailbox ?

**Reply by JorjMcKie** [reply]
I have opened issue #1244 to address this problem correctly.

**Reply by p1-dta** [reply]
Do you have any idea about when the fix will be released for get_image_bbox ?

**Reply by JorjMcKie** [reply]
I am working on it. Should be no later than tomorrow.
